### PR TITLE
RDoc-871 Add working-with-document-ids, hilo-algorithm and handling-case-sensitive-document-ids

### PR DESCRIPTION
**Pull request by IdanHaim**

### CODE REVIEW

File: Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+{INFO:Prefix convention}
+Note that we used `companies/` as the prefix just to follow the RavenDB convention. However nothing stands in the way to provide the different prefix, 
+which will be completely unrelated to the collection name.
+{INFO/}
+
+{INFO:Concurrent writes}
+The identities are generated and updated on the server side in the atomic fashion. This means you can safely use this approach in the concurrent writes scenario.
+{INFO/}
+
+## Commands usage
```
Comment (ppekrol) [inline Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown]: I'm wondering if we need this section? Thoughts?

File: Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/handling-case-sensitive-document-ids.markdown [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# Handling case sensitive document IDs
+
+RavenDB document IDs are handled by design not to be case sensitive. Load operations will execute regardless of the letter casing of the IDs.
```
[the remark this answers is not in the document]
Comment (IdanHaim) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/handling-case-sensitive-document-ids.markdown]: done

File: Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/handling-case-sensitive-document-ids.markdown [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+The trick to handle this case is encode the case sensitive key into a unique representation that can be stored in a case insensitive way.
```
[the remark this answers is not in the document]
Comment (IdanHaim) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/handling-case-sensitive-document-ids.markdown]: The context is how to handle case senstive Id's in ravendb what you couldn't understand.
In 3.5 article we just explain why you need it but something is wrong and I got confuse from it.
client that want to know how to handle a case sensetive document IDs will know why he need it

File: Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/hilo-algorithm.markdown [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# HiLo algorithm
+
+The HiLo algorithm is an efficient solution used by [a session](../session/what-is-a-session-and-how-does-it-work) to generate numeric parts of identifiers. In other words it is
+responsible for providing numeric values that are combined with collection names and node ID to create keys like `orders/10-A`, `products/93-B` etc. 
+
+The client is able to determine to what type of the collection does an entity belong. In order to differ documents, it also needs to add a unique number and the node tag at the end of the document's key.
```
[the remark this answers is not in the document]
Comment (IdanHaim) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/hilo-algorithm.markdown]: done

File: Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/hilo-algorithm.markdown [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+To ensure that multiple clients can generate the keys simultaneously, they need some mechanism to avoid duplicates. It is ensured with `Replication/HiLo/collection` documents, stored
```
[the remark this answers is not in the document]
Comment (IdanHaim) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/hilo-algorithm.markdown]: done

File: Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown [hunk 6]
```diff
@@ -0,0 +1,131 @@
+# Working with document identifiers
+
+Each document in RavenDB database has a unique string associated with this document, called an identifier or a key. Every entity that you store either by using [a session](../session/what-is-a-session-and-how-does-it-work)
+or with a [put doucment command](../commands/documents/put) will have such identifier assigned in the database. In RavenDB there are [four options](../../server/kb/document-key-generation) supported by the server to store a document and assign an identifier to it.
```
[the remark this answers is not in the document]
Comment (IdanHaim) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown]: I don't think is too low level and if soo I am giving an example on how to do it with the command .
I think removing this line we make the example to be out of the blue. don't you ?

[the remark this answers is not in the document]
Comment (IdanHaim) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown]: About the key We can change it to ID maybe beacuse I am using it in the aritcle

File: Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+## Session usage
+
+If you choose to use the session, you don't have to pay any special attention to the identifiers of the stored entities. The session will take care of it by generating the keys automatically.
+It utilizes [conventions](../../client-api/configuration/conventions/what-are-conventions) and HiLo algorithm to produce the identifiers. Everything is handled by the session's mechanism and is transparent for the user. 
+However, you can influence the key generation strategy by overwriting [the key generation conventions](../../client-api/configuration/conventions/identifier-generation/global). In this article we are going to consider the behavior 
+in accordance with the default conventions.
+
+{INFO:Identifiers in RavenDB are strings}
+Identifiers of documents in RavenDB database are always strings. RavenDB is designed to work with string identifiers and that is the best option you should choose
```
[the remark this answers is not in the document]
Comment (IdanHaim) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown]: rephrase

File: Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+After execute the code above we will get from the server ID that looks like this `companies/000000000000000027-A`.
+Be aware that the only guarantee for the auto IDs is that it will always increase in the same node.
+
+Using `|` at the end of the Id is quite the same as putting `/`, in this case we are going to tell RavenDB to create
+the ID when the document is saved but here we are going to use a speciel cluster identity integer that will increase everytime we are using the pipe symbol
+as a suffix to our Id.
+
+{CODE session_identity_id@ClientApi\DocumentIdentifiers\WorkingWithDocumentIds.cs /}
+
+After execute the code above the Id will be `companies/1`, here we don't add the tag of the node to the end of the ID
```
[the remark this answers is not in the document]
Comment (IdanHaim) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown]: done

File: Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+{INFO/}
+
+
+### Autogenerated ids
+
+In order to figure out which property (or field) holds the entity's identifier, the convention `Conventions.FindIdentityProperty` is called. 
+By default, it looks for the property or the field named `Id` (case sensitive). However this property can have `null` value or even not be present at all. Then the automatic key
+generation strategy is performed. The default convention is that such entities get the keys in the following format `collection/number-tag`. RavenDB client first determines
+the name of [the collection](../../client-api/faq/what-is-a-collection) that the entity belongs to, and then contacts with the server in order to retrieve a numeric range of values, which can be used as the `number` part.
+The range of available numbers is calculated by using `HiLo` algorithm and it is tracked per collection. The current maximum value in ranges are stored in documents `Raven/Hilo/collectionName`.
```
[the remark this answers is not in the document]
Comment (IdanHaim) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown]: done

File: Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown (hunk 1 above)
Comment (IdanHaim) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown]: I think we need to explain different why tou can achive what you need. I don't know if some client use only the commands but if they do they can see it here

File: Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+as a suffix to our Id.
+
+{CODE session_identity_id@ClientApi\DocumentIdentifiers\WorkingWithDocumentIds.cs /}
+
+After execute the code above the Id will be `companies/1`, here we don't add the tag of the node to the end of the ID
+because this number is unique to all the cluster.
+Identities are sequential, so running the above code again will generate `companies/2`, and so on.
+
+{INFO:Prefix convention}
+Note that we used `companies/` as the prefix just to follow the RavenDB convention. However nothing stands in the way to provide the different prefix,
```
[the remark this answers is not in the document]
Comment (IdanHaim) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown]: Yes it is importent will add

File: Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/handling-case-sensitive-document-ids.markdown (hunk 3 above)
[the remark this answers is not in the document]
Comment (ayende) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/handling-case-sensitive-document-ids.markdown]: I think we can drop this entirely. That isn't an interesting thing to talk about.

**Comment by IdanHaim** [comment]
I think you are right, I will remove it


On Jan 4, 2018 11:02, "Ayende Rahien" <notifications@github.com> wrote:

*@ayende* commented on this pull request.
------------------------------

In Documentation/4.0/Raven.Documentation.Pages/client-api/
document-identifiers/handling-case-sensitive-document-ids.markdown
<https://github.com/ravendb/docs/pull/436#discussion_r159604352>:

> @@ -0,0 +1,14 @@
+# Handling case sensitive document IDs
+
+RavenDB document IDs are handled by design not to be case sensitive.
Load operations will execute regardless of the letter casing of the
IDs.
+The trick to handle this case is encode the case sensitive key into a
unique representation that can be stored in a case insensitive way.

I think we can drop this entirely. That isn't an interesting thing to talk
about.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/ravendb/docs/pull/436#discussion_r159604352>, or mute
the thread

File: Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown (hunk 6 above)
[the remark this answers is not in the document]
Comment (ppekrol) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown]: yes, you have 'identifier' there already so no need to add ID

[the remark this answers is not in the document]
Comment (IdanHaim) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/working-with-document-ids.markdown]: Done

File: Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/handling-case-sensitive-document-ids.markdown (hunk 3 above)
[the remark this answers is not in the document]
Comment (IdanHaim) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/handling-case-sensitive-document-ids.markdown]: done

File: Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/hilo-algorithm.markdown [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+}
+{CODE-BLOCK/}
+
+The `Max` property means the maximum possible number that has been used by any client to create the key for a given collection. It is used as follows:
+
+1. The client ask the server for a range of numbers that he can use to create document (32 is the default)
+2. then the server check the hilo file to see what is the last number he sent to any client for this collection.
+3. The client will get from the server the min and the max values he can use (33 - 64 in our case).
+4. Then the client generate a range object from the values he got from the server to generate identifiers.
+5. When the client reach the max limit he will repeat the process until he done.
```
[the remark this answers is not in the document]
Comment (IdanHaim) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/document-identifiers/hilo-algorithm.markdown]: done

**Pull request merged**